mono/mini/debugger-agent.c

```diff
@@ -1677,30 +1677,55 @@ transport_handshake (void)
 }
 
 static void
-stop_debugger_thread (void)
+wait_for_debugger_thread_to_stop()
 {
-	if (!agent_inited)
-		return;
-
-	transport_close1 ();
-
-	/* 
+	/*
 	 * Wait for the thread to exit.
 	 *
 	 * If we continue with the shutdown without waiting for it, then the client might
 	 * not receive an answer to its last command like a resume.
 	 */
-	if (!is_debugger_thread ()) {
+	if (!is_debugger_thread()) {
 		do {
-			mono_coop_mutex_lock (&debugger_thread_exited_mutex);
+			mono_coop_mutex_lock(&debugger_thread_exited_mutex);
 			if (!debugger_thread_exited)
-				mono_coop_cond_wait (&debugger_thread_exited_cond, &debugger_thread_exited_mutex);
-			mono_coop_mutex_unlock (&debugger_thread_exited_mutex);
+				mono_coop_cond_wait(&debugger_thread_exited_cond, &debugger_thread_exited_mutex);
+			mono_coop_mutex_unlock(&debugger_thread_exited_mutex);
 		} while (!debugger_thread_exited);
 
 		if (debugger_thread_handle)
-			mono_thread_info_wait_one_handle (debugger_thread_handle, MONO_INFINITE_WAIT, TRUE);
+			mono_thread_info_wait_one_handle(debugger_thread_handle, MONO_INFINITE_WAIT, TRUE);
 	}
+}
+
+static void
+stop_debugger_thread (void)
+{
+	if (!agent_inited)
+		return;
+
+#ifdef HOST_WIN32
+	gboolean debuggerAttached = mono_is_debugger_attached();
+
+	// We need to make the call to mono_threads_suspend_abort_syscall to break any
+	// hung accept calls on windows.
+	MonoThreadInfo* info = mono_thread_info_lookup(debugger_thread_id);
+	if (info) {
+		mono_threads_suspend_abort_syscall(info);
+
+		// On Windows we must wait for the debugger to stop before
+		// closing the transport. !debuggerAttached means we are not in
+		// socket accept, the problematic blocking call.
+		if (!debuggerAttached)
+		{
+			wait_for_debugger_thread_to_stop();
```

Comment:
It looks like we are going to call this below no matter what, so is the check here useful?

Reply:
Previously the closing of the socket would trigger the `debugger_thread` to end if a debugger was attached and then we'd wait for the thread to stop. We wanted to keep that behavior as well as handling the case where the debugger wasn't attached and the `accept` call the debugger thread was blocking on needed to be cancelled.